### ISSUE TITLE
Great variance in number of dropped epochs

### ISSUE THREAD

**Issue by SebastianSpeer** [autoreject/autoreject#180]
Dear Autoreject Team,

I am currently preprocessing some data from a Stroop task and have noticed that when running autoreject on the same subject several times the number of bad epochs varies considerably from 11 bad epochs dropped up to 38 bad epochs dropped (out of 72 epochs in total). Do you have some suggestions as to where this variance is combing from and ss there a way to reduce this variance?

Thanks,

Sebastian

**Comment by dengemann** [comment]
Hi! It sounds like a random seed is not set. Which is surprising and points potentially at a bug. Can you provide a minimum reproducible example with toy data ?

**Comment by SebastianSpeer** [comment]
[Test-AR.zip](https://github.com/autoreject/autoreject/files/4722860/Test-AR.zip)

```
decision_epochs = mne.read_epochs(opj(out_dir,task,'Test-AR.fif'), preload=True)
ar = AutoReject(verbose='tqdm_notebook', n_jobs=n_jobs )
decision_epochs_ar, rejection_log_fix = ar.fit_transform(decision_epochs, return_log=True)
```

**Comment by jasmainak** [comment]
can you provide a full script which we can run by copy-pasting?

and also two seeds -- one where few epochs are dropped and one where too many epochs are dropped. Thanks.

**Comment by SebastianSpeer** [comment]
```
import os
from os.path import join as opj
import mne
import numpy as np
import scipy.io as sio
import matplotlib.pyplot as plt
from pyprep.noisy import Noisydata
from autoreject import AutoReject
from autoreject import get_rejection_threshold
from autoreject import Ransac
import time

start_time = time.time()

cutoff_l=1
cutoff_h=100
downsample=500
n_jobs=3
manu_reject=160e-6
reference='average'

def ensure_dir(ed):
    try: 
        os.makedirs(ed)
    except OSError:
        if not os.path.isdir(ed):
            raise
# Load data
# do not exclude  channel named 'status' because it's the stimulus channel
raw=mne.io.read_raw_bdf('Test_AR_raw.bdf', eog=['EXG1','EXG2','EXG3','EXG4'], exclude=['EXG5', 'EXG6','EXG7','EXG8','GSR1','GSR2','Erg1','Erg2','Resp','Plet','Temp'] ,preload=True) 

# Add a montage to the data
montage =mne.channels.make_standard_montage("biosemi64")
raw.set_montage(montage)

# Extract some info
eeg_index = mne.pick_types(raw.info, eeg=True, eog=False, meg=False)
ch_names = raw.info["ch_names"]
ch_names_eeg = list(np.asarray(ch_names)[eeg_index])
sample_rate = raw.info["sfreq"]
task_duration=max(raw.times)/60



#--- Find events

events = mne.find_events(raw, initial_event=True)

if 111 not in events[:,2]:
    # some strange error in the events so need to subtract 256 from all events
    events[:,2]=events[:,2]-256 
events_list=list(events[:,2])


new_events=mne.merge_events(events,[111,112,113,114],110, replace_events=True)
new_events=mne.merge_events(new_events,[121,122,123,124],120, replace_events=True)
new_events=mne.merge_events(new_events,[131,132,133,134],130, replace_events=True)
new_events=mne.merge_events(new_events,[141,142,143,144],140, replace_events=True)

# Filtering

raw_highpass = raw.copy().filter(l_freq=cutoff_l, h_freq=cutoff_h)


# line noise removal

freqs=np.arange(50, sample_rate/2, 50)
raw_notch = raw_highpass.copy().notch_filter(freqs=freqs)


# Downsampling

raw_downsampled = raw_notch.copy().resample(sfreq=downsample)


# Find noisy channels

nd = Noisydata(raw_downsampled)
nd.find_all_bads(ransac=False)
bads = nd.get_bads(verbose=True)


decision_events_of_interest = {
    'Congruent/Corresponding': 110,
    'Incongurent/Corresponding': 120,
    'Congruent/NotCorresponding': 130,
    'Incongruent/NotCorresponding': 140}
    #'Button1':88, 'Button2':89,'TargetButton1':98,'TargetButton2':99}

decision_epochs = mne.Epochs(raw_downsampled, new_events, tmin=-0.4, tmax=0.5,baseline=(-0.3,0),
                    #reject=reject_criteria,  flat=flat_criteria, # commented beause we're using autoreject here
                    event_id = decision_events_of_interest,
                    reject_by_annotation=False, preload=True)

# find more noisy channels using ransac and interpolate them
# this is done separately to allow using several jobs
ransac=Ransac(n_resample=30,n_jobs=n_jobs)
ransac.fit(decision_epochs)
bads=ransac.bad_chs_ + list(set(bads) - set(ransac.bad_chs_))

print(bads)



# add the bad channel info 
raw_downsampled.info['bads']=bads

print("---  Identifying noisy channels done: %s seconds ---" % (time.time() - start_time))

# Interpolating bad channels

eeg_data_interp = raw_downsampled.copy().interpolate_bads(reset_bads=True)

print("---  Interpolating bad channels done: %s seconds ---" % (time.time() - start_time))



# use the average of all channels as reference

if reference=='average':
    eeg_data_interp_ref = eeg_data_interp.copy().set_eeg_reference(ref_channels='average')

print("---  referencing done: %s seconds ---" % (time.time() - start_time))


#-- ICA --- #

ica = mne.preprocessing.ICA(n_components=40, random_state=1, allow_ref_meg=True)


# -----------------------------#


ica.fit(eeg_data_interp_ref)

ica.exclude = []
# find which ICs match the EOG pattern
eog_indices, eog_scores = ica.find_bads_eog(eeg_data_interp_ref, ch_name='EXG1')

ica.exclude = eog_indices

# plot ICs applied to raw data, with EOG matches highlighted
fig=ica.plot_sources(eeg_data_interp_ref)


eeg_data_interp_ref_ica = eeg_data_interp_ref.copy()
ica.apply(eeg_data_interp_ref_ica)

print("---  ICA done: %s seconds ---" % (time.time() - start_time))

#---------- Epoching -------------#

#-- short ---#
reject_criteria = dict(eeg=manu_reject)       # 150 µV - Should be set as input
flat_criteria = dict(eeg=1e-6) 


decision_epochs_short = mne.Epochs(raw_downsampled, new_events, tmin=-0.3, tmax=0.5,baseline=(-0.3,0),
                    #reject=reject_criteria,  flat=flat_criteria, # commented beause we're using autoreject here
                    event_id = decision_events_of_interest,
                    reject_by_annotation=False, preload=True)



# autoreject producing 10 dropped epochs
ar = AutoReject(verbose='tqdm_notebook', n_jobs=n_jobs, random_state=30 )
decision_epochs_ar, rejection_log_fix = ar.fit_transform(decision_epochs, return_log=True)


# autoreject producing 35 dropped epochs
ar = AutoReject(verbose='tqdm_notebook', n_jobs=n_jobs, random_state=35 )
decision_epochs_ar, rejection_log_fix = ar.fit_transform(decision_epochs, return_log=True)
```

**Comment by SebastianSpeer** [comment]
turns out I can't upload the raw bdf which I'm loading in the script here because it's too big. But you can just use the file from my previous comment which corresponds to the data after epoching in the second script. Or I can send the file in another way

**Comment by eioe** [comment]
Hi,
I'd like to second @SebastianSpeer 's question.  
I observe a similar behavior of AR in my current data set - with different seeds producing substantial differences in the number of rejected epochs. In my case it's 720 epochs per subject and if I remember correctly, for some of them the results varied between ~30 up to ~120 rejected epochs by AR when re-running with a different seed (I haven't systematically assessed this yet.). If needed, I can try to post/share data + a minimal example, but @SebastianSpeer already did so and I don't want to hijack the issue.

**Comment by jasmainak** [comment]
hi @eioe and @SebastianSpeer 

I'm sorry I haven't gotten around to this yet. I'll have to sit down with @dengemann or @agramfort and investigate. I doubt it is a bug though. I think it arises from the rho and kappa parameters, what we call `consensus_perc` and `n_interpolate`. If you have a grid of `n_interpolate` values that is a bit too sparse, e.g., `[1, 2, 4, 32]` you might end up changing the results by a lot when changing random seeds if the minima is shallow (e.g., in Figure 1 of the paper). You could probably test this hypothesis  by making the grid more dense (but it will be painfully slow). Or if you're lazy like me, you can fix them to a fixed value manually. The thresholds will still be found automatically but you turn off the hyperparameter selection for the more unstable part. I don't know if this is true but I haven't found the time to investigate :-(

**Comment by guishuyunye-lyw** [comment]
> hi @eioe and @SebastianSpeer
> 
> I'm sorry I haven't gotten around to this yet. I'll have to sit down with @dengemann or @agramfort and investigate. I doubt it is a bug though. I think it arises from the rho and kappa parameters, what we call `consensus_perc` and `n_interpolate`. If you have a grid of `n_interpolate` values _that is a bit too sparse_, e.g., `[1, 2, 4, 32]` you might end up changing the results by a lot when changing random seeds if the minima is shallow (e.g., in Figure 1 of the paper). You could probably test this hypothesis by making the _grid more dense_ (but it will be painfully slow). Or if you're lazy like me, you can fix them to a _fixed value manually_. The thresholds will still be found automatically but you turn off the hyperparameter selection for the more unstable part. I don't know if this is true but I haven't found the time to investigate :-(

hi @jasmainak ,What's the fixed value you recommend. I found you set the consensus_perc = 0.6, which means only a majority of channel marked as bad, the trial will be drop.     Is it too high?

**Comment by guishuyunye-lyw** [comment]
> ```
> import os
> from os.path import join as opj
> import mne
> import numpy as np
> import scipy.io as sio
> import matplotlib.pyplot as plt
> from pyprep.noisy import Noisydata
> from autoreject import AutoReject
> from autoreject import get_rejection_threshold
> from autoreject import Ransac
> import time
> 
> start_time = time.time()
> 
> cutoff_l=1
> cutoff_h=100
> downsample=500
> n_jobs=3
> manu_reject=160e-6
> reference='average'
> 
> def ensure_dir(ed):
>     try: 
>         os.makedirs(ed)
>     except OSError:
>         if not os.path.isdir(ed):
>             raise
> # Load data
> # do not exclude  channel named 'status' because it's the stimulus channel
> raw=mne.io.read_raw_bdf('Test_AR_raw.bdf', eog=['EXG1','EXG2','EXG3','EXG4'], exclude=['EXG5', 'EXG6','EXG7','EXG8','GSR1','GSR2','Erg1','Erg2','Resp','Plet','Temp'] ,preload=True) 
> 
> # Add a montage to the data
> montage =mne.channels.make_standard_montage("biosemi64")
> raw.set_montage(montage)
> 
> # Extract some info
> eeg_index = mne.pick_types(raw.info, eeg=True, eog=False, meg=False)
> ch_names = raw.info["ch_names"]
> ch_names_eeg = list(np.asarray(ch_names)[eeg_index])
> sample_rate = raw.info["sfreq"]
> task_duration=max(raw.times)/60
> 
> 
> 
> #--- Find events
> 
> events = mne.find_events(raw, initial_event=True)
> 
> if 111 not in events[:,2]:
>     # some strange error in the events so need to subtract 256 from all events
>     events[:,2]=events[:,2]-256 
> events_list=list(events[:,2])
> 
> 
> new_events=mne.merge_events(events,[111,112,113,114],110, replace_events=True)
> new_events=mne.merge_events(new_events,[121,122,123,124],120, replace_events=True)
> new_events=mne.merge_events(new_events,[131,132,133,134],130, replace_events=True)
> new_events=mne.merge_events(new_events,[141,142,143,144],140, replace_events=True)
> 
> # Filtering
> 
> raw_highpass = raw.copy().filter(l_freq=cutoff_l, h_freq=cutoff_h)
> 
> 
> _# line noise removal_
> 
> _freqs=np.arange(50, sample_rate/2, 50)_
> raw_notch = raw_highpass.copy().notch_filter(freqs=freqs)
> 
> 
> # Downsampling
> 
> raw_downsampled = raw_notch.copy().resample(sfreq=downsample)
> 
> 
> # Find noisy channels
> 
> nd = Noisydata(raw_downsampled)
> nd.find_all_bads(ransac=False)
> bads = nd.get_bads(verbose=True)
> 
> 
> decision_events_of_interest = {
>     'Congruent/Corresponding': 110,
>     'Incongurent/Corresponding': 120,
>     'Congruent/NotCorresponding': 130,
>     'Incongruent/NotCorresponding': 140}
>     #'Button1':88, 'Button2':89,'TargetButton1':98,'TargetButton2':99}
> 
> decision_epochs = mne.Epochs(raw_downsampled, new_events, tmin=-0.4, tmax=0.5,baseline=(-0.3,0),
>                     #reject=reject_criteria,  flat=flat_criteria, # commented beause we're using autoreject here
>                     event_id = decision_events_of_interest,
>                     reject_by_annotation=False, preload=True)
> 
> # find more noisy channels using ransac and interpolate them
> # this is done separately to allow using several jobs
> ransac=Ransac(n_resample=30,n_jobs=n_jobs)
> ransac.fit(decision_epochs)
> bads=ransac.bad_chs_ + list(set(bads) - set(ransac.bad_chs_))
> 
> print(bads)
> 
> 
> 
> # add the bad channel info 
> raw_downsampled.info['bads']=bads
> 
> print("---  Identifying noisy channels done: %s seconds ---" % (time.time() - start_time))
> 
> # Interpolating bad channels
> 
> eeg_data_interp = raw_downsampled.copy().interpolate_bads(reset_bads=True)
> 
> print("---  Interpolating bad channels done: %s seconds ---" % (time.time() - start_time))
> 
> 
> 
> # use the average of all channels as reference
> 
> _if reference=='average':
>     eeg_data_interp_ref = eeg_data_interp.copy().set_eeg_reference(ref_channels='average')1_
> 
> print("---  referencing done: %s seconds ---" % (time.time() - start_time))
> 
> 
> #-- ICA --- #
> 
> ica = mne.preprocessing.ICA(n_components=40, random_state=1, allow_ref_meg=True)
> 
> 
> # -----------------------------#
> 
> 
> ica.fit(eeg_data_interp_ref)
> 
> ica.exclude = []
> # find which ICs match the EOG pattern
> eog_indices, eog_scores = ica.find_bads_eog(eeg_data_interp_ref, ch_name='EXG1')
> 
> ica.exclude = eog_indices
> 
> # plot ICs applied to raw data, with EOG matches highlighted
> fig=ica.plot_sources(eeg_data_interp_ref)
> 
> 
> eeg_data_interp_ref_ica = eeg_data_interp_ref.copy()
> ica.apply(eeg_data_interp_ref_ica)
> 
> print("---  ICA done: %s seconds ---" % (time.time() - start_time))
> 
> #---------- Epoching -------------#
> 
> #-- short ---#
> reject_criteria = dict(eeg=manu_reject)       # 150 µV - Should be set as input
> flat_criteria = dict(eeg=1e-6) 
> 
> 
> decision_epochs_short = mne.Epochs(raw_downsampled, new_events, tmin=-0.3, tmax=0.5,baseline=(-0.3,0),
>                     #reject=reject_criteria,  flat=flat_criteria, # commented beause we're using autoreject here
>                     event_id = decision_events_of_interest,
>                     reject_by_annotation=False, preload=True)
> 
> 
> 
> # autoreject producing 10 dropped epochs
> ar = AutoReject(verbose='tqdm_notebook', n_jobs=n_jobs, random_state=30 )
> decision_epochs_ar, rejection_log_fix = ar.fit_transform(decision_epochs, return_log=True)
> 
> 
> # autoreject producing 35 dropped epochs
> ar = AutoReject(verbose='tqdm_notebook', n_jobs=n_jobs, random_state=35 )
> decision_epochs_ar, rejection_log_fix = ar.fit_transform(decision_epochs, return_log=True)
> ```

To ignore bad segments using autoreject (local), we could do:

>>> ar = AutoReject()
>>> _, reject_log = ar.fit(epochs).transform(epochs, return_log=True)
>>> ica.fit(epochs[~reject_log.bad_epochs])

hi @SebastianSpeer @jasmainak , what do you think about the process, which combine the autoreject  and ICA together.  I got some confusion about it, after ica.fit(), ica.apply,should I run fit_transform() again?

**Comment by jasmainak** [comment]
> I found you set the consensus_perc = 0.6, which means only a majority of channel marked as bad, the trial will be drop. Is it too high?

It sounds like a reasonable number. Do make sure to check your data visually afterwards

**Comment by guishuyunye-lyw** [comment]
> > I found you set the consensus_perc = 0.6, which means only a majority of channel marked as bad, the trial will be drop. Is it too high?
> 
> _It sounds like a reasonable number. Do make sure to check your data visually afterwards_

I thought 0.6 is a littler higher. Is there any recommended criteria?

**Comment by jasmainak** [comment]
I would look at the data.

**Comment by AleKrugliak** [comment]
Hello Autoreject Team,

I am new to Python and EEG analysis. I am struggling with the same issue of great variance of dropped epochs. The ideas for potential solutions mentioned above are not very clear to me. Would you mind giving step for step instructions how I can circumvent the issue?

Many thanks in advance!

Best wishes,
Ale

**Comment by jasmainak** [comment]
I'm sorry I haven't found time to get around to this issue. 

Simple solution for now is fix the seed and inspect what autoreject gives you
